treeppl/tree-inference-naive.tppl

```diff
@@ -0,0 +1,311 @@
+ * 
+ * @param Q rate matrix
+ * 
+ * @return a matrix encoding the transition probabilities
+ *           from one state to another 
+ *
+ */
+transition_probabilities = function(Q) => Probability [4, 3]
+{
+    P = [ [-Q[0][1]/Q[0][0], -Q[0][2]/Q[0][0], -Q[0][3]/Q[0][0] ],
+	  [-Q[1][0]/Q[1][1], -Q[1][2]/Q[1][1], -Q[1][3]/Q[1][1] ],
+	  [-Q[2][0]/Q[2][2], -Q[2][1]/Q[2][2], -Q[2][3]/Q[2][2] ],
+	  [-Q[3][0]/Q[3][3], -Q[3][1]/Q[3][3], -Q[3][2]/Q[3][3] ] ]
+    
+    return P
+}
+
+
+
+/**
+ * CTMC Process for a sequence of states
+ *
+ */
+ctmc = function(
+    parental_state: State,
+    evolution_time: Time,
+    Q: InstantaneousRateMaterix,
+    P: Probability[4][3],
+    mu: PositiiveReal = 1
+) => State
+{
+   
+    possible_next_states = States[-parental_state]
+    
+    // The waiting time until the transition is given by the negative
+    // of the diagonal of the transition matrix
+    lambda = -mu*Q[parental_state][parental_state]
+    t ~ Exponential(rate = lambda)
+    
+    if (evolution_time - t < 0) {
+	// no transition
+	return parental_state
+    }
+
+    new_state ~ Categorical(probabilities =  P[starting_state], support = possible_next_states)
+    
+    return evolve(new_state, evolution_time - t, Q, P, mu)
+}
+
+
+/** 
+ * Stationary distribution of the transition matrix Q
+ *  Amounts to calculating the left eigenvector of the matrix
+ *  Hint:  https://brilliant.org/wiki/stationary-distributions/
+ */
+stationary_distribution = function(Q: InstantaneousRateMatrix) => Probability[4]
+{
+    // TODO implement stationary distribution
+}
+
+
+
+/// Tree Stuff ///
+
+/** Leaf 
+ * @param index the leaf number
+ * @param states the states of the leaf
+ * @return leaf
+ */
+leaf = function(index: BigInt, states: State[]) => Leaf
+{
+    return {
+	type: Leaf,
+	index: index,
+	age: 0.0,
+	states: states
+    }
+}
+
+
+/** Node */
+node = function(left: Tree | Leaf, right: Tree | Leaf, age: NonnegativeReal, states: State[]) => Tree
+{
+    return {
+	type: Node,
+	left: left,
+	right: right,
+	age: age,
+	states: states
+    }
+}
+
+/** Convert tree to Newick */
+top_to_newick = function(t: Tree)
+{
+    // TODO not implemented
+}
+
+/// Core model functions ///
+
+
+/** Recursive function to randomly build a tree */
+cluster = function(
+    Q: InstantaneousRateMatrix,
+    trees: Array<Tree>,
+    max_age : Age
+) => Array<Tree>
+{
+    // Check if we are done
+    if trees.length == 1
+        return trees
+
+    // Pick two random trees in the trees vector
+    v ~ Permutation(0:(trees.length - 1))
+    i = v[0]
+    j = v[1]
+
+    // Propose s (new node states) from arbitrary proposal distribution (here stationary distribution)
+    propose s ~ iid(
+	dist = Categorical(probabilities = stationary_distribution(Q), support = States),
+	n = trees[i].states.length
+    )
+    
+    // Propose node_age (new node age) from prior 
+    t ~ Exponential(rate = 20.0)
+    parental_age = max_age + t 
```

Comment:
Using capital letter here suggests to me that Exponential is a recognized type, and Exponential() is the function that constructs an object of this type. Correct?